File: backend/app/routers/cash_account.py
"""CashAccount router — REST endpoints for cash balance management."""

from __future__ import annotations

from fastapi import APIRouter, Path, Query, status

from app.core.deps import CashAccountServiceDep, CurrentUser
from app.models.cash_account_transaction import CashTxKind
from app.schemas.auth import ErrorResponse
from app.schemas.cash_account import (
    CashAccountCreate,
    CashAccountResponse,
    CashAccountTransactionResponse,
    CashAccountUpdate,
)

router = APIRouter(prefix="/api/cash-accounts", tags=["cash-accounts"])


@router.get(
    "",
    response_model=list[CashAccountResponse],
    status_code=status.HTTP_200_OK,
    summary="List all cash accounts",
    responses={
        401: {"model": ErrorResponse, "description": "Not authenticated"},
    },
)
async def list_cash_accounts(
    _current_user: CurrentUser,
    service: CashAccountServiceDep,
) -> list[CashAccountResponse]:
    """Return all cash accounts ordered by creation date descending."""
    accounts = await service.list()
    return [CashAccountResponse.model_validate(a) for a in accounts]


@router.get(
    "/transactions",
    response_model=list[CashAccountTransactionResponse],
    status_code=status.HTTP_200_OK,
    summary="List cash account transactions (interest, deposits, etc.)",
    responses={
        401: {"model": ErrorResponse, "description": "Not authenticated"},
    },
)
async def list_cash_account_transactions(
    _current_user: CurrentUser,
    service: CashAccountServiceDep,
    kind: CashTxKind | None = Query(
        default=None,
        description="Optional kind filter (e.g. ``interest`` to return only interest payments)",
    ),
) -> list[CashAccountTransactionResponse]:
    """Return cash account transactions newest-first; optional kind filter."""
    txs = await service.list_transactions(kind=kind)
    return [CashAccountTransactionResponse.model_validate(t) for t in txs]


@router.post(
    "",
    response_model=CashAccountResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new cash account",
    responses={
        401: {"model": ErrorResponse, "description": "Not authenticated"},
        422: {"model": ErrorResponse, "description": "Validation error"},
    },
)
async def create_cash_account(
    data: CashAccountCreate,
    _current_user: CurrentUser,
    service: CashAccountServiceDep,
) -> CashAccountResponse:
    """Create a cash account with the specified label, currency, and balance."""
    account = await service.create(data)
    return CashAccountResponse.model_validate(account)


@router.patch(
    "/{id}",
    response_model=CashAccountResponse,
    status_code=status.HTTP_200_OK,
    summary="Partially update a cash account",
    responses={
        401: {"model": ErrorResponse, "description": "Not authenticated"},
        404: {"model": ErrorResponse, "description": "Cash account not found"},
        422: {"model": ErrorResponse, "description": "Validation error"},
    },
)
async def update_cash_account(
    data: CashAccountUpdate,
    _current_user: CurrentUser,
    service: CashAccountServiceDep,
    id: int = Path(..., ge=1, description="Cash account ID to update"),
) -> CashAccountResponse:
    """Update label and/or balance of an existing cash account.

    Returns 404 if the cash account does not exist.
    Returns 422 if neither label nor balance is provided, or if currency is sent.
    """
    account = await service.update(id, data)
    return CashAccountResponse.model_validate(account)


@router.delete(
    "/{id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a cash account",
    responses={
        401: {"model": ErrorResponse, "description": "Not authenticated"},
        404: {"model": ErrorResponse, "description": "Cash account not found"},
    },
)
async def delete_cash_account(
    _current_user: CurrentUser,
    service: CashAccountServiceDep,
    id: int = Path(..., ge=1, description="Cash account ID to delete"),
) -> None:
    """Hard-delete a cash account by id.

    Returns 404 if the cash account does not exist.
    """
    await service.delete(id)

File: frontend/src/lib/api/cash-account.ts
import { apiClient } from "@/lib/api-client";
import type {
  CashAccount,
  CashAccountTransaction,
  CashTxKind,
} from "@/types/cash-account";
import type {
  CashAccountCreateInput,
  CashAccountUpdateInput,
} from "@/lib/schemas/cash-account";

// ── Raw shape (snake_case from backend) ──────────────────────────────────────

interface RawCashAccount {
  id: number;
  label: string;
  currency: string;
  balance: string;
  created_at: string;
  updated_at: string;
}

// ── Converter ─────────────────────────────────────────────────────────────────

function toCashAccount(raw: RawCashAccount): CashAccount {
  return {
    id: raw.id,
    label: raw.label,
    currency: raw.currency,
    balance: raw.balance,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

// ── Public API helpers ─────────────────────────────────────────────────────────

export async function getCashAccounts(): Promise<CashAccount[]> {
  const response = await apiClient.get<RawCashAccount[]>("/api/cash-accounts");
  return response.data.map(toCashAccount);
}

export async function createCashAccount(
  input: CashAccountCreateInput,
): Promise<CashAccount> {
  const response = await apiClient.post<RawCashAccount>(
    "/api/cash-accounts",
    input,
  );
  return toCashAccount(response.data);
}

export async function updateCashAccount(
  id: number,
  input: CashAccountUpdateInput,
): Promise<CashAccount> {
  const response = await apiClient.patch<RawCashAccount>(
    `/api/cash-accounts/${id}`,
    input,
  );
  return toCashAccount(response.data);
}

export async function deleteCashAccount(id: number): Promise<void> {
  await apiClient.delete(`/api/cash-accounts/${id}`);
}

interface RawCashTransaction {
  id: number;
  cash_account_id: number | null;
  kind: CashTxKind;
  amount: string;
  currency: string;
  traded_at: string;
  external_source: string | null;
}

function toCashTransaction(raw: RawCashTransaction): CashAccountTransaction {
  return {
    id: raw.id,
    cashAccountId: raw.cash_account_id,
    kind: raw.kind,
    amount: raw.amount,
    currency: raw.currency,
    tradedAt: raw.traded_at,
    externalSource: raw.external_source,
  };
}

export async function listCashTransactions(params: {
  kind?: CashTxKind;
} = {}): Promise<CashAccountTransaction[]> {
  const response = await apiClient.get<RawCashTransaction[]>(
    "/api/cash-accounts/transactions",
    { params: { kind: params.kind } },
  );
  return response.data.map(toCashTransaction);
}
